### PR TITLE
Add test c-host errata search commands web ui

### PR DESCRIPTION
**Pull request by swadeley**
Tests search commands in web UI's Content Hosts page.
Tests for Applicable versus Installable

closed loop for:
[Bug 1707335 - Unable to search the content host list in Satellite 6.4 webui using the filter 'errata_status = security_needed' and 'errata_status = errata_needed'](https://bugzilla.redhat.com/show_bug.cgi?id=1707335)

### CODE REVIEW

**Comment by swadeley** [comment]
robottelo]$ pytest --show-capture=no tests/foreman/ui/test_errata.py -k test_content_host_errata_search_commands
====================== test session starts ===================================
platform linux -- Python 3.7.7, pytest-4.6.3, py-1.8.1, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /sat6/robottelo
plugins: cov-2.8.1, services-1.3.1, mock-1.10.4, forked-1.1.3, xdist-1.31.0
collecting ... 2020-06-06 21:24:32 - conftest - DEBUG - Collected 14 test cases
2020-06-06 22:24:32 - robottelo.issue_handlers.bugzilla - DEBUG - Calling Bugzilla API for {'1655130'}
2020-06-06 22:24:35 - robottelo.ssh - DEBUG - Instantiated Paramiko client 0x7f9c626691d0
2020-06-06 22:24:35 - robottelo.ssh - INFO - Connected to [dell-r330-12.xxxx.redhat.com]
2020-06-06 22:24:35 - robottelo.ssh - INFO - >>> rpm -q satellite
2020-06-06 22:24:37 - robottelo.ssh - INFO - <<< stdout
satellite-6.8.0-0.4.beta.el7sat.noarch

2020-06-06 22:24:37 - robottelo.ssh - DEBUG - Destroyed Paramiko client 0x7f9c626691d0
2020-06-06 22:24:37 - robottelo.host_info - DEBUG - Host Satellite version: 6.8
2020-06-06 22:24:37 - robottelo.helpers - DEBUG - Generated file bz_cache.json with BZ collect data
collected 14 items / 13 deselected / 1 selected                                                                  

tests/foreman/ui/test_errata.py .                                                                          [100%]

===================== 1 passed, 13 deselected in 794.65 seconds ===================

File: tests/foreman/ui/test_errata.py [hunk 1]
```diff
@@ -757,3 +762,97 @@ def test_positive_filtered_errata_status_installable_param(session, errata_statu
...
+
+@tier3
+def test_content_host_errata_search_commands(session, module_org, module_repos_col):
+    """ View a list of affected content hosts for security (RHSA) and bugfix (RHBA) errata,
+    filtered with errata status and applicable flags. Applicability is calculated using the
+    Library, but Installability is calculated using the attached CV, and is subject to the
+    CV's own filtering.
+
+:id: 45114f8e-0fc8-4c7c-85e0-f9b613530dac
+
```
Comment (pondrejk) [inline tests/foreman/ui/test_errata.py:775]: we should probably mention the BZ in the docstring

Comment (swadeley) [reply tests/foreman/ui/test_errata.py:775]: oh yes, will do

Comment (swadeley) [reply tests/foreman/ui/test_errata.py:775]: Hello @pondrejk 
I can see :BZ: 1707335 at the end of the docstring. Is that the wrong place?

Comment (swadeley) [reply tests/foreman/ui/test_errata.py:775]: @pondrejk I can see it at the end of the docstring, is that incorrect place?

Comment (swadeley) [reply tests/foreman/ui/test_errata.py:775]: Searching other bugs I see :BZ: typically comes after ID or expectedresults, but sometimes before and sometimes after after CaseLevel or CaseImportance (in the later case, that is often at the end).

Comment (pondrejk) [reply tests/foreman/ui/test_errata.py:775]: I think the order of items in docstring is not significant

Comment (mirekdlugosz) [reply tests/foreman/ui/test_errata.py:775]: @swadeley Peter is correct, order of metadata doesn't matter

File: tests/foreman/ui/test_errata.py [hunk 2]
```diff
@@ -757,3 +762,97 @@ def test_positive_filtered_errata_status_installable_param(session, errata_statu
...
+:Steps:
+    1.  host list --search "errata_status = security_needed"
+    2.  host list --search "errata_status = errata_needed"
+    3.  host list --search "applicable_errata = RHSA-2012:0055"
+    4.  host list --search "applicable_errata = RHBA-2012:1030"
+    5.  host list --search "applicable_rpms = walrus-5.21-1.noarch"
+    6.  host list --search "applicable_rpms = kangaroo-0.3-1.noarch"
+    7.  Create filter & rule to hide RPM (Applicable vs. Installable test)
+    8.  Repeat steps 3 and 5, but 5 expects host name not found.
```
Comment (mirekdlugosz) [inline tests/foreman/ui/test_errata.py:786]: Is this relevant to bugzilla at hand?
Don't we cover similar scenarios elsewhere?

File: tests/foreman/ui/test_errata.py (hunk 1 above)
Comment (swadeley) [reply tests/foreman/ui/test_errata.py:775]: In that case, there is nothing wrong here, BZ: 1707335  is present in docstring.

**Comment by swadeley** [comment]
Hello @mirekdlugosz 

> I don't understand. Referenced bugzilla is still in NEW, but your test finishes with pass?
the bug is against 6.4, unfortunately I cannot get the test to run against a 6.4 Satellite after changing to the 6.4 branch due to:
`INTERNALERROR> pluggy.manager.PluginValidationError: unknown hook 'pytest_namespace' in plugin`

I could try create a new venv just for Sat6.4 tests. i.e. change to the 6.4 branch, then create the venv, then run pip install

> Does that mean that once bugzilla is fixed, our test will start to fail?
No, the fact that it passes on 6.8 is not too surprising as the bug was against 6.4
> Or once bugzilla is fixed this test will continue to pass, which might mean that we don't really verify issue at hand?
See above on troubles testing against 6.4, any tips on how to run test using old branch of robottelo? The git branch does not keep your properties file and venv setup.

Thank you

**Comment by swadeley** [comment]
I noticed this test does not work on 6.7.1, yet it works for 6.8

Consider these two lines (846 - 847):

    `# Get the LCE created last
    lce = content_view.environment[-1]`

=================================================
    This is on a 6.7.1 Satellite
     
    nailgun.entities.LifecycleEnvironment(id=6)                                        
    >>> print(content_view_version)                                                    
    nailgun.entities.ContentViewVersion(content_view=nailgun.entities.ContentView(i  
    d=7), description=None,                                                          
    environment=[nailgun.entities.LifecycleEnvironment(id=6)],                    
    errata_counts={'security': 13, 'bugfix': 9, 'enhancement': 2, 'total': 24},        
    file_count=0, major=2, minor=0, module_stream_count=0, package_count=126,    
    puppet_module=[], repository=[nailgun.entities.Repository(id=43),              
    nailgun.entities.Repository(id=44), nailgun.entities.Repository(id=45)],        
    version='2.0', id=10)                                                          
    >>> print(content_view.environment)                                            
    [nailgun.entities.LifecycleEnvironment(id=7),      <--- thats the one we want, the last one created                    
    nailgun.entities.LifecycleEnvironment(id=6)]                                    
    >>> print(content_view.environment[-1])                                        
    nailgun.entities.LifecycleEnvironment(id=6)        <--- Not that one, as I was expecting by using [-1]                
    >>> print(content_view.environment[0])                                          
    nailgun.entities.LifecycleEnvironment(id=7)   <--- thats the one we want
     
     
    =================================================
     
    This is on a 6.8 Satellite
     
    >>> print(content_view.environment)                                              
    [nailgun.entities.LifecycleEnvironment(id=7),                                    
    nailgun.entities.LifecycleEnvironment(id=8)]        <--- thats the one we want            
    >>> print(content_view.environment[-1])                                          
    nailgun.entities.LifecycleEnvironment(id=8)   <--- thats the one we want, so this worked as I expected in 6.8, but not 6.7.1                                
    >>> print(content_view.environment[0])                                      
    nailgun.entities.LifecycleEnvironment(id=7)       <--- Not that one        
    >>> print(content_view.environment[1])                                          
    nailgun.entities.LifecycleEnvironment(id=8)

File: tests/foreman/ui/test_errata.py (hunk 2 above)
Comment (swadeley) [reply tests/foreman/ui/test_errata.py:786]: I can remove the bugzilla number if you prefer?
 I cannot find any test that checks Applicable vs. Installable in the sense of an erratum is not installable, because its not in the CV, but it is still shown as Applicable (its in the Library and the host needs it).

Comment (swadeley) [reply tests/foreman/ui/test_errata.py:786]: @mirekdlugosz , I did add this test in CLI section recently: https://github.com/SatelliteQE/robottelo/blob/9b80440f0df4662f5383f278a8d759b666b7fff9/tests/foreman/cli/test_errata.py#L578

So I can remove those last two steps from this test.

**Comment by JacobCallahan** [comment]
@swadeley you could always iterate through the environments and pull out the one with the highest id. Or you could sort the list by id's and use [-1]

**Comment by swadeley** [comment]
> @swadeley you could always iterate through the environments and pull out the one with the highest id. Or you could sort the list by id's and use [-1]

thanks @JacobCallahan , this seems to work:
`lce = sorted(content_view.environment, key=id)[-1] `

**Comment by swadeley** [comment]
Hello @mirekdlugosz , I have removed the last section that did the applicability versus installable check.

Please review and merge.

Thank you

robottelo]$ pytest --show-capture=no tests/foreman/ui/test_errata.py -k test_content_host_errata_search_commands
=============================== test session starts ==================================
platform linux -- Python 3.7.7, pytest-4.6.3, py-1.8.1, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /sat6/robottelo
plugins: cov-2.8.1, services-1.3.1, mock-1.10.4, forked-1.1.3, xdist-1.31.0
collecting ... 2020-07-10 09:23:20 - conftest - DEBUG - Collected 14 test cases
2020-07-10 10:23:20 - robottelo.issue_handlers.bugzilla - DEBUG - Calling Bugzilla API for {'1655130'}
2020-07-10 10:23:22 - robottelo.ssh - DEBUG - Instantiated Paramiko client 0x7ff1b11da890
2020-07-10 10:23:22 - robottelo.ssh - INFO - Connected to [dell-r330-12.xxxx.redhat.com]
2020-07-10 10:23:22 - robottelo.ssh - INFO - >>> rpm -q satellite
2020-07-10 10:23:25 - robottelo.ssh - INFO - <<< stdout
satellite-6.8.0-0.6.beta.el7sat.noarch

2020-07-10 10:23:25 - robottelo.ssh - DEBUG - Destroyed Paramiko client 0x7ff1b11da890
2020-07-10 10:23:25 - robottelo.host_info - DEBUG - Host Satellite version: 6.8
2020-07-10 10:23:25 - robottelo.helpers - DEBUG - Generated file bz_cache.json with BZ collect data
collected 14 items / 13 deselected / 1 selected                                                                  

tests/foreman/ui/test_errata.py .                                                                          [100%]

======================== 1 passed, 13 deselected in 865.69 seconds =====================

File: tests/foreman/ui/test_errata.py (hunk 2 above)
Comment (mirekdlugosz) [reply tests/foreman/ui/test_errata.py:784]: @swadeley I think you have removed this filtering part from test itself, so it's only matter of updating steps in docstring now.

File: tests/foreman/ui/test_errata.py [hunk 3]
```diff
@@ -757,3 +760,74 @@ def test_positive_filtered_errata_status_installable_param(session, errata_statu
...
+
+
+@tier3
+def test_content_host_errata_search_commands(session, module_org, module_repos_col):
+    """ View a list of affected content hosts for security (RHSA) and bugfix (RHBA) errata,
+    filtered with errata status and applicable flags. Applicability is calculated using the
+    Library, but Installability is calculated using the attached CV, and is subject to the
+    CV's own filtering.
+
+:id: 45114f8e-0fc8-4c7c-85e0-f9b613530dac
```
Comment (mirekdlugosz) [inline tests/foreman/ui/test_errata.py:772]: Haven't noticed that before, but metadata (i.e. everything down to end of docstring, including closing `"""`) should be indented by one level.

File: tests/foreman/ui/test_errata.py [hunk 4]
```diff
@@ -757,3 +760,74 @@ def test_positive_filtered_errata_status_installable_param(session, errata_statu
...
+            assert client2.hostname in result
+            # Search for RHSA applicable RPMs
+            result = session.contenthost.search(f'applicable_rpms = {FAKE_2_CUSTOM_PACKAGE}')
+            result = [item['Name'] for item in result]
+            assert client1.hostname in result
+            # Search for RHBA applicable RPMs
+            result = session.contenthost.search(f'applicable_rpms = {FAKE_5_CUSTOM_PACKAGE}')
+            result = [item['Name'] for item in result]
+            assert client2.hostname in result
+            # Search for installable RHSA errata by Errata ID
```
Comment (mirekdlugosz) [inline tests/foreman/ui/test_errata.py:826]: This and next one seem to not be included in steps.

File: tests/foreman/ui/test_errata.py (hunk 2 above)
Comment (swadeley) [reply tests/foreman/ui/test_errata.py]: Right, thanks.

File: tests/foreman/ui/test_errata.py (hunk 4 above)
Comment (swadeley) [reply tests/foreman/ui/test_errata.py:825]: Well spotted, thank you, changes now made.

File: tests/foreman/ui/test_errata.py (hunk 3 above)
Comment (swadeley) [reply tests/foreman/ui/test_errata.py]: Yes, fixed now, thank you

**Comment by swadeley** [comment]
Hello

testing blocked by [Bug 1855348 - katello_applicability accidentally set to true at install](https://bugzilla.redhat.com/show_bug.cgi?id=1855348)

**Comment by mirekdlugosz** [comment]
@swadeley I believe you can modify file in question on your instance, restart Satellite and try this test out. Of course it will fail on automation run.
You want to add quick `skip_if_open` decorator and merge, or hold this until bz is resolved?

**Comment by swadeley** [comment]
@mirekdlugosz 

I did try edit answer file:

`[root@dell-r330-12 ~]# grep yum /etc/foreman-installer/scenarios.d/satellite-answers.yaml
  enable_yum: true
  proxy_pulp_yum_to_pulpcore: true
  enable_yum: true
  use_pulp_2_for_yum: true`

then ran the installer, checked katello file:

    `[root@dell-r330-12 ~]# grep -r -A5 use_pulp_2_for_content_type /etc/foreman/plugins/katello.yaml
      :use_pulp_2_for_content_type:
        :docker: true
        :file: true
        :yum: true

      :katello_applicability: false`

but I did not try a restart of services (I think installer does restart services).

 Lets just wait for next snap rather than the use skip_if_open and then get nagged to reverse that so soon. I will add do not merge flag.

Thank you

**Comment by swadeley** [comment]
Hello @mirekdlugosz 

Testing on latest snap

robottelo]$ git checkout host_errata_search_commands_web_ui
Switched to branch 'host_errata_search_commands_web_ui'
robottelo]$ pytest --show-capture=no tests/foreman/ui/test_errata.py -k test_content_host_errata_search_commands
============================= test session starts =================================
platform linux -- Python 3.7.7, pytest-4.6.3, py-1.8.1, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /sat6/robottelo
plugins: cov-2.8.1, services-1.3.1, mock-1.10.4, forked-1.1.3, xdist-1.31.0
collecting ... 2020-07-21 13:56:04 - conftest - DEBUG - Collected 14 test cases
2020-07-21 14:56:04 - robottelo.issue_handlers.bugzilla - DEBUG - Calling Bugzilla API for {'1655130'}
2020-07-21 14:56:06 - robottelo.ssh - DEBUG - Instantiated Paramiko client 0x7fcac5538a50
2020-07-21 14:56:06 - robottelo.ssh - INFO - Connected to [dhcp-3-238.xxxx.redhat.com]
2020-07-21 14:56:06 - robottelo.ssh - INFO - >>> rpm -q satellite
2020-07-21 14:56:08 - robottelo.ssh - INFO - <<< stdout
satellite-6.8.0-0.7.beta.el7sat.noarch

2020-07-21 14:56:08 - robottelo.ssh - DEBUG - Destroyed Paramiko client 0x7fcac5538a50
2020-07-21 14:56:08 - robottelo.host_info - DEBUG - Host Satellite version: 6.8
2020-07-21 14:56:08 - robottelo.helpers - DEBUG - Generated file bz_cache.json with BZ collect data
collected 14 items / 13 deselected / 1 selected                                                                  

tests/foreman/ui/test_errata.py .                                                                          [100%]

======================= 1 passed, 13 deselected in 704.46 seconds ==============

**Comment by swadeley** [comment]
oops, merge conflict, I will push fix and submit new test result

**Comment by swadeley** [comment]
robottelo]$ git status
On branch host_errata_search_commands_web_ui
nothing to commit, working tree clean
robottelo]$ pytest --show-capture=no tests/foreman/ui/test_errata.py -k test_content_host_errata_search_commands
==================================== test session starts ==========================================
platform linux -- Python 3.7.7, pytest-4.6.3, py-1.8.1, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /sat6/robottelo
plugins: cov-2.8.1, services-1.3.1, mock-1.10.4, forked-1.1.3, xdist-1.31.0
collecting ... 2020-07-21 18:21:59 - conftest - DEBUG - Collected 14 test cases
collected 14 items / 13 deselected / 1 selected                                                                  

tests/foreman/ui/test_errata.py .                                                                          [100%]

============================= 1 passed, 13 deselected in 677.06 seconds ==============================

**Comment by swadeley** [comment]
Thank you all.
